### ISSUE TITLE
Loading more than one tile async

### ISSUE THREAD

**Issue by bnsantos** [moagrius/TileView#250]
Hey,
I did a PR while ago when the lib was still in version 1.something to fetch more than one tile at the same time. And it seems that this code was not introduced in version 2.0, all tiles are loaded in sequence (takes too long to load all tiles if they're not as a resource inside the app). 

Do you have interest in updating your lib to load more than one tile at the same time? I could possibly work on that and create another PR but before I would like to know why do you remove the code I did for this (was there a bug) not in the scope of the lib?

Thanks,

**Comment by moagrius** [comment]
Hi - I remember the update.  I don't remember specifically how the final threading model was decided on, but virtually _everything_ in version 2 was reviewed and rewritten.  I would guess it's that the current iteration uses a `ThreadPoolExecutor`, so it's not really one at a time, it's just ordered - I believe that it'll start as many operations as there are cores available, but yes they'd be executed in order (and might potentially block subsequent ops; I'd have to look it up and am only half-in right now (still at work)).

Do you believe your approach would demonstrat better actual (versus perceptual) performance?  Can you briefly describe (or remind me) what your alternate approach was?  Unfortunately I squashed and lost history with v2.

**Comment by bnsantos** [comment]
Hey,

I'm just getting familiar with your whole new code. But it seems that is exactly the same point as the previous versoin. You have a ThreadPoolExecutor that usually only have one async task which is responsible for loading all tiles sequentially (TileRenderTask in TileCanvasViewGroup.java). 

A faster approach would be having each tile render in a separate Thread this way the ThreadPoolExecutor could run more than one Thread at the same time. But in a approach like that we need to have a better way to decide if the task should run or not, because every time you scroll or do something else you should cancel all previous queued tasks and queue new tasks for each tile (you just need to cancel one task in your code right now and create a new instance of TileRenderTask).

I worked in this multi thread approach because I'm using the TileView with assets on the Internet, being able to download more than one at the same time is a must. I could possibly work in a way to make Tile loading multi threaded and do a PR for you to review.

It seems that my fork is still in the same state as the PR I did, if you want to take a look
https://github.com/bnsantos/TileView

**Comment by moagrius** [comment]
Thanks for the link - that does look familiar.

The `AsyncTask` for fetching tile bitmaps in v1 was _not_ using `executeOnExecutor(THREAD_POOL_EXECUTOR)`, but v2 does.  

AFAICT, using `AsyncTask` and passing in the static `ThreadPoolExecutor` _will_ use multiple threads - one per available core (see the third link)

https://github.com/android/platform_frameworks_base/blob/master/core/java/android/os/AsyncTask.java#L202

Am I'm misreading that?

**Comment by bnsantos** [comment]
Hey Mike,

You're not wrong about the ThreadPoolExecutor API. If you have more than one Tasks/Thread it will be able to run multiple tasks at the same time depending on the number of the cores available. But in v2 you have only one TileRenderTask per time in the ThreadPoolExecutor. This way doesn't matter if you have 1, 2 or N cores ThreadPoolExecutor will have only one Task/Thread.

Every time the user stops span, zomming or first open the TileView the TileRenderTask is responsible for loading all tiles in a sequence (only starts to load/render a tile after the previous tile was successfully loaded rendered). And that TileRenderTask goes to ThreadPoolExecutor and lies there until either it finishes or some zoom or span happens. In case of zoom or span happens you add a new TileRenderTask to load the tile in the new position while the previous task is being invalidate (it will be removed eventually from ThreadPoolExecutor not sure when exactly) there will be more than one Task in the ThreadPoolExecutor (only time that will happen).

To properly use MultiThread you need to spawn a lot of Threads/AsyncTask in the ThreadPoolExecutor. So it can run more than one at a time depending on the number of cores available.

To make it more clear I recorded two videos of my app using TileView. One with the version v1 and the other with the version v2. The images come from a website and they were not cached previously. I ran in the same emulator both versions.
[Tile Loading v1 - MultiThread](https://www.dropbox.com/s/pbiroq80yz80pky/multi_threaded.mp4?dl=0)
[Tile Loading v2 - SingleThread](https://www.dropbox.com/s/fmh9y5y9zsyc8g5/single_threaded.mp4?dl=0)
Even though the Tiles are loaded in sequence for v1 and v2 you can see how fast v1 is. It is fast because it is downloading more than one image at the same time because I spawn a TileDownloadRunner for every Tile.

I hope you understand my points

**Comment by moagrius** [comment]
makes perfect sense, i'm embarrassed to have not known that - yes, by all means PR - or if you don't have the time, i can do it based on your original work - let me know what works best for you.

if you do decide to create the update yourself, please try to match style and convention (naming, whitespace, class structure, etc) as close to existing as possible.

thanks again - i appreciate your work and cluing me in onto what i'd misunderstood

**Comment by bnsantos** [comment]
Good,
I'll work on those changes and keep the same style convention!
Thanks

**Comment by peterLaurence** [comment]
Hi bnsantos,

Are you actually working on this? I would like to give it a try, but if you have already started, i guess there is no point doing the same thing twice ;)

**Comment by bnsantos** [comment]
Hey  @peterLaurence ,

No I didn't even started this, I was thinking of starting this right now. Did you start already?

Sorry for the delayed reply I was pulled into some urgent stuff here and I didn't even have time to read your message.

**Comment by peterLaurence** [comment]
Hi @bnsantos ,

No problem. I planned to do it but i don't have time right now. As you may have seen, some problem have been fixed the last couple of weeks, including a critical one (see #260).
Look forward to see what you hearing whay you come up with.  If you make it work with `tileView.shouldRenderWhilePaning( true )`, it would be perfect.

**Comment by bnsantos** [comment]
Hey @peterLaurence 
Just finished a "working" version in my [fork](https://github.com/bnsantos/TileView) if you could please take a look because you're more up to date on those issues. I'm still trying to catch up the issues we're open while I was absent. 

Also added an example to download tiles from internet.

**Comment by peterLaurence** [comment]
Ok i will try that!

**Comment by peterLaurence** [comment]
Looks very good! It works with `shouldRenderWhilePaning( true )` (LargeImage example).
And your example works nice too.

**Comment by peterLaurence** [comment]
Will take a closer look at your code as soon as i have time. (i hope tomorrow). In the meantime, you may have to comply with Mike's code style :)

**Comment by moagrius** [comment]
@bnsantos yes please try to conform to existing code style, whitespace, naming conventions, patterns, etc.  if you could re-commit when that's done and post back, i/we can do a quick code review and try to get this integrated asap

thanks again

**Comment by peterLaurence** [comment]
So i did some checking. Even if the overall performance of this new implementation of tile loading is very good, i was curious to know if it is better than the old one for local tiles (ofc this new implementation is better for tiles on the net). So basically i defined this in `TileView` to log the render time :
```
private long time;

@Override
public void onRenderStart() {
  time = System.currentTimeMillis();
}

  
@Override
public void onRenderComplete() {
  long delta = System.currentTimeMillis() - time;
  System.out.println( "delta : " + delta );
}
```

Then i picked the LargeImage demo example and i double-taped several times to go through all `DetailLevel` in a cyclic manner. On my nexus 5 (2013 edition), those are delta with the old implementation :

02-12 23:28:37.120 11590-11590/tileview.demo I/System.out: delta : 196
02-12 23:28:39.514 11590-11590/tileview.demo I/System.out: delta : 1
02-12 23:28:39.580 11590-11590/tileview.demo I/System.out: delta : 0
02-12 23:28:41.555 11590-11590/tileview.demo I/System.out: delta : 267
02-12 23:28:42.694 11590-11590/tileview.demo I/System.out: delta : 1
02-12 23:28:42.705 11590-11590/tileview.demo I/System.out: delta : 1
02-12 23:28:42.721 11590-11590/tileview.demo I/System.out: delta : 0
02-12 23:28:42.759 11590-11590/tileview.demo I/System.out: delta : 0
02-12 23:28:43.312 11590-11590/tileview.demo I/System.out: delta : 230
02-12 23:28:47.266 11590-11590/tileview.demo I/System.out: delta : 1
02-12 23:28:47.310 11590-11590/tileview.demo I/System.out: delta : 1
02-12 23:28:50.015 11590-11590/tileview.demo I/System.out: delta : 133
02-12 23:29:00.913 11590-11590/tileview.demo I/System.out: delta : 176

I believe the 0 and the 1 are caused by a `requestRender` directly followed by a `cancelRender`, which happens at a `DetailLevel` change.


And with the new implementation (one thread per tile) :

02-12 23:21:05.947 6090-6090/tileview.demo I/System.out: delta : 135
02-12 23:21:07.155 6090-6090/tileview.demo I/System.out: delta : 164
02-12 23:21:10.288 6090-6090/tileview.demo I/System.out: delta : 229
02-12 23:21:10.473 6090-6090/tileview.demo I/System.out: delta : 138
02-12 23:21:12.557 6090-6090/tileview.demo I/System.out: delta : 196
02-12 23:21:16.723 6090-6090/tileview.demo I/System.out: delta : 214
02-12 23:21:16.886 6090-6090/tileview.demo I/System.out: delta : 101
02-12 23:21:18.753 6090-6090/tileview.demo I/System.out: delta : 146
02-12 23:21:21.516 6090-6090/tileview.demo I/System.out: delta : 237
02-12 23:21:21.719 6090-6090/tileview.demo I/System.out: delta : 157

In conclusion render times are very close, perhaps slightly better for the new method. Maybe you can test with your own devices.

**Comment by peterLaurence** [comment]
@bnsantos 

About your code, while i like how the `Runnable` post their result on the handler which runs on ui thread, it seems like you assume that the last `TileRenderRunnable` you create in `TileRenderPoolExecutor.queue` will be the last to end, thus it calls `TileCanvasViewGroup.onRenderTaskPostExecute`. It seems to behave like that, but if the last `TileRenderRunnable` finishes before one of the other launched before, the call of `TileCanvasViewGroup.onRenderTaskPostExecute` which then calls a `TileCanvasViewGroup.cleanup`. That last call a little too early may cause an issue but i am not sure of that.

**Comment by bnsantos** [comment]
@moagrius 
I'm aware of your new coding style because of that I already refactored a lot of think, and before the PR I would ask you to review again.

@peterLaurence 
The render time should be the same because I didn't change the `BitmapProvider` implementation (in the example you mentioned `BitmapProviderAssets`) what I'm doing is a way to render more than one Tile per time if the device has more than one core (can even load 3, 4, ... 8 at the same time). So I would suggest you to add another log 

```
@Override
public void onRenderStart() {
  time = System.currentTimeMillis();
  System.out.println( "Start loading tile : " + <some tile identifier here> );
}

@Override
public void onRenderComplete() {
  long delta = System.currentTimeMillis() - time;
  System.out.println( "Finished rendering tile: "+ <some tile identifier here>  + " - delta : " + delta );
}
```

Then you could see if more than one tile is rendering at the same time.

Now regarding the last `TileRenderRunnable` being queued that I add a flag saying that this is the last one to be **queued** but not necessary it will be the last one to be executed. You're right! I missed that . Going to take a look to see how can I fix that to make sure to call `TileCanvasViewGroup.onRenderTaskPostExecute` after all queued `TileRenderRunnable` are executed. Thanks

**Comment by moagrius** [comment]
> I'm aware of your new coding style because of that I already refactored a lot of think, and before the PR I would ask you to review again.

:thumbsup:

**Comment by peterLaurence** [comment]
@bnsantos 

I know that the same `BitmapProvider` used by each thread is the same as before. But the whole task of running threads versus one thread loading all tiles can differ in execution time, depending on how concurrency is handled at lower levels and how the overhead of switching between threads affects performance.
For tiles on a network, thread are mostly IO-bound so having several threads is definitely better (maybe more than the number of cores could be better).
For local tiles, we are mostly CPU-bound, and then having a thread equal to number of cores is good compromise.

As for the test i made, i'd like to clarify that `TileView.onRenderStart()` is called once before the first tile render, and `TileView.onRenderComplete()` is called after the last tile render. So the delta between the two calls should measure the execution time of the whole rendering task. That's what i wanted to benchmark.
That's reassuring though that those execution time are equivalent.

**Comment by bnsantos** [comment]
@peterLaurence 
My bad, got confused between `Tile` and `TileView`. You're right about the loading time. And no worries with switching threads because according the [documentation] (https://developer.android.com/reference/java/util/concurrent/ThreadPoolExecutor.html) 

> Thread pools address two different problems: they usually provide improved performance when executing large numbers of asynchronous tasks, due to reduced per-task invocation overhead, and they provide a means of bounding and managing the resources, including threads, consumed when executing a collection of tasks. Each ThreadPoolExecutor also maintains some basic statistics, such as the number of completed tasks.

So I think it's finally done, now I have a custom implementation of ThreadPoolExecutor that has a hooker method `afterExecute` that checks if it is the last running thread.

**Comment by peterLaurence** [comment]
@bnsantos 

:+1: Looks good to me

**Comment by moagrius** [comment]
if i'm looking at the right branch (master) looks like there's still some style issues (mostly whitespace: https://github.com/moagrius/TileView/compare/master...bnsantos:master)...  did you want to PR and we can do line comments, or do you want to fix it up a bit first?

**Comment by bnsantos** [comment]
@moagrius 
So I changed more coding style stuff. Now I'll create the PR and we can do line comments.
Thanks
